feature-crowdloan-impl/src/main/java/jp/co/soramitsu/feature_crowdloan_impl/presentation/contribute/confirm/ConfirmContributeViewModel.kt

```diff
@@ -144,38 +169,95 @@ class ConfirmContributeViewModel(
+                        }
+                        payload.metadata.isMoonbeam && ethAddress?.second == true -> {
+                            additionalOnChainSubmission(it, flowName, payload.amount, customContributeManager)
+                        }
+                        payload.metadata.isAcala && payload.contributionType == 1 -> {
+                            additionalOnChainSubmission(it, flowName, payload.amount, customContributeManager)
+                        }
+                        else -> {
+                            null
+                        }
+                    }
                 }
 
-                contributionInteractor.contribute(
-                    originAddress = selectedAddressModelFlow.first().address,
-                    parachainId = payload.paraId,
-                    contribution = payload.amount,
-                    token = assetFlow.first().token,
-                    additionalSubmission
-                )
+                val isLcDotAcala = payload.metadata?.isAcala == true && payload.contributionType == 1
+                if (isLcDotAcala) {
+                    val recipient = contributionInteractor.getAcalaStatement(payload.metadata?.flow?.data?.baseUrl!!).proxyAddress
+                    val maxAllowedStatusLevel = if (suppressWarnings) TransferValidityLevel.Warning else TransferValidityLevel.Ok
+                    val fee = feeFlow.firstOrNull()?.feeModel?.fee ?: return@launch
+                    contributionInteractor.performTransfer(
+                        transfer = Transfer(
+                            recipient = recipient,
+                            amount = payload.amount,
+                            tokenType = Token.Type.DOT
+                        ),
+                        fee = fee,
+                        maxAllowedLevel = maxAllowedStatusLevel,
+                        additional = additionalSubmission
+                    )
+                } else {
+                    contributionInteractor.contribute(
+                        originAddress = selectedAddressModelFlow.first().address,
+                        parachainId = payload.paraId,
+                        contribution = payload.amount,
+                        token = assetFlow.first().token,
+                        additionalSubmission,
+                        privateCrowdloanSignature
+                    )
+                }
             }
                 .onFailure(::showError)
                 .onSuccess {
                     showMessage(resourceManager.getString(R.string.common_transaction_submitted))
 
+                    saveMoonbeamEtheriumAddress()
+
                     router.returnToMain()
                 }
 
             _showNextProgress.value = false
         }
     }
+
+    fun warningConfirmed() {
+        sendTransaction(suppressWarnings = true)
+    }
+
+    private suspend fun saveMoonbeamEtheriumAddress() {
+        if (payload.metadata?.isMoonbeam == true) {
```

Comment:
Please consider short return instead of wrapping whole function with extra right padding

Reply:
But better do when(...)